### PR TITLE
Netfilter

### PR DESCRIPTION
**Pull request by tombentley**
This *draft* PR is not yet in a working state, but should be advanced enough for some initial feedback. 

The basic moving parts here are as follows:
* A `PROXYHandler` for dealing with the HAProxy PROXY protocol "connection header", so that kproxy can make use of the client source address when fronted by HAProxy (this could be split into a separate PR).
* A `KafkaAuthnHandler` for performing SASL authentication offload (i.e. the proxy does the authn). This is reusing some of Kafka's own SASL code and I've implemented support for SASL/Plain and SASL/SCRAM-SHA-256 and SASL/SCRAM-SHA-512. We can add support for OAUTHBEARER (and even GSSAPI) at a later stage.
* The `NetHandler` support the `NetFilter` SPI, which provides an extension point for selecting which upstream cluster the client should be connected to, and with which `KrpcFilters`. The `NetFilter.NetFilterContext` provides information, such as sniHostname, or authorized id, to a `NetFilter` implementation which has been obtained from prior handlers in the pipeline (optionally any of Netty's own `SSLHandler` (for SNI host), `PROXYHandler` and `KafkaAuthnHandler`) via "user events"
* Using that info, the `NetFilter` implementation is expected to call `NetFilterContext.connect()` exactly once. In other words the incoming "messages" could traverse any one of the following paths, depending on how things are configured@
    ```
    ─╮─ SSLHandler ─╭─╮─ PROXYHandler ─╭─╮─ KafkaAuthnHandler ─╭─── NetHandler
     ╰──────────────╯ ╰────────────────╯ ╰─────────────────────╯ 
    ```
* The `SimpleNetFilter` replicates the same functionality we already had: Connecting to a single statically known cluster with a statically known chain of `KrpcFilters`.
* Obviously you could support a "gateway" proxy which connected to different clusters depending on SNI hostname and/or authorised id. It could also use a different filter chain based on these.
* Further applications would include tar pitting, and closing connections based on this information (this is why propagating source addresses it useful, since we could make those decisions in part based on client IP address)

---

Apart from the general untidiness plus whatever other comments you raise, the main problem yet to resolve is that (in a pipeline with just the `NetHandler`, no SSL, PROXY or authn offload):
1. a producer makes the ApiVersions request
2. the NetHandler returns the ApiVersions response
3. the producer makes the Metadata request
4. the NetHandler passes control to the SimpleNetFilter, which calls back on `NetHandler.connect`, we create a `KafkaProxyFrontendHandler` and add it to the pipeline, but it doesn't get registered or activated by Netty. We don't have a Context for it in order to do those things ourselves.
5. When we subsequently call `ctx.fireChannelRead(msg);` with the `MetadataRequestData` (in `NetHandler.channelRead`) the metadata request gets silently dropped.
6. The subsequent `InitProduerId` request from the Producer does make it to the broker and the response returned, but it now has the wrong sequence number and the Producer throws because it never saw a response to the `Metadata` request.

### CODE REVIEW

**Comment by tombentley** [comment]
> The subsequent InitProduerId request from the Producer does make it to the broker

Maybe it's due to the async connect in the FrontentHandler, but an initial experiment with waiting for that didn't seem to work out.

**Comment by tombentley** [comment]
Oh also in here are `BareSasl(Request|Response)` which we'd need to support in the codec, but would allow to support older clients that don't use `SaslAuthenticate`.

File: pom.xml [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
     <properties>
         <java.version>11</java.version>
-        <java.test.version>17</java.test.version>
+        <java.test.version>11</java.test.version>
```
Comment (gunnarmorling) [inline pom.xml:21]: What's the reason for going back here? I suppose it will cause some existing tests to fail.

Comment (tombentley) [reply pom.xml:21]: IntelliJ didn't respect the different JVM for test as for the main code. So it was seeing parser/compiler errors on the multi-line string literals and refusing to run the tests.

**Comment by tombentley** [comment]
> we create a KafkaProxyFrontendHandler and add it to the pipeline, but it doesn't get registered or activated by Netty. We don't have a Context for it in order to do those things ourselves.

I've dug into this a little. We can lookup the context for a handler that's been added to the pipeline from the pipeline itself. But whether or not we use that context to manually activate the added handler ourselves (which feels wrong) the problem is then that the FrontendHandler is not yet able to handle the request because the BackendHandler is not yet activated, and so the flow control mechanism prevents it.

Overall I think I'm missing some insight about channel activation (and possibly the asynchrony of that) that I hope @franz1981 can provide.

File: pom.xml (hunk 1 above)
Comment (gunnarmorling) [reply pom.xml:21]: I see. So indeed in IDEs, the configuration needs to be set to 17 (as most (all?) of them don't differentiate between different levels for main and test code). This was a once-off config for me in Eclipse (and probably I'll have to redo it when updating the project config). That felt acceptable to me, given the advantages of e.g. being able to specify YAML in test code inline, without any escaping or line concatenation. If we want to change it to 11, some existing tests will need to be adjusted. My personal vote would be on keeping 11 for main and 17 for testing.

Comment (tombentley) [reply pom.xml:21]: Let me see if I can make [IDEA-85478](https://youtrack.jetbrains.com/issue/IDEA-85478) work

Comment (tombentley) [reply pom.xml:21]: With some fiddling that seems to work. Had to switch to using the maven wrapper, rather than IDEA's built-in maven (which was too old a version) and also tick "Create separate modules for production and test roots" in `Settings > Maven > Importing`, and finally configure the right JVM for the test.

**Comment by tombentley** [comment]
This now works. Summary of recent work:

* Merged `NetFilterHandler` in `KafkaProxyFrontendHandler`, which results in much simpler code (due to not having to insert a fronetndhandler into the channel pipeline while handling an in-flight read in the net handler). This means the currently the frontend handler _always_ handles apiversions itself, even if every other kind of KRPC request gets sent to a backend cluster. 
* Keep the authn handler more or less unchanged. This means that the proxy can optionally handle kafka sasl authn itself, or pass it though depending on whether this handler is configured. This has tests, but I've not yet wired it up into the `KafkaProxy` part, so you can't actually use it yet. 
* Removed `PROXYHandler` (since netty has its own support for this), but keep support for consuming the haproxy message and feeding it into the net filter. 

Prior to merging we probably need tests for:

* The state machine in `KafkaProxyFrontendHandler`
* An integration test for the autn offload functionality
* Nice to have: A test for the haproxy functionality. Integration-style test is probably painful to write currently, but maybe something possible using an embedded channel?

However, I think this is much much closer to the final form of what a net filter might look like, so PTAL!

**Comment by tombentley** [comment]
@franz1981 ptal

**Comment by gunnarmorling** [comment]
> A PROXYHandler for dealing with the HAProxy PROXY protocol "connection header", so that kproxy can make use of the client source address when fronted by HAProxy (this could be split into a separate PR).

Would it be a lot of work to do that split? If not, I'd be +1 on separating it, to make things more digestable here. But if it is, let's keep it as one.

**Review by gunnarmorling — COMMENTED** [review]
Some first small comments after a quick review pass. Only on details, sorry for commenting on those first. Will do add more after a second pass once I've better groked the larger picture.

**Comment by tombentley** [comment]
> Would it be a lot of work to do that split?

I don't think so. I'll try to have a look next week.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/frame/BareSaslResponse.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.frame;
+
+/**
+ * Ancient versions of Kafka implemented SASL/GSSAPI by sending the response
+ * on the wire as length prefixed bytes (no Kafka protocol header).
+ * This frame represents those kinds of response.
+ */
+public class BareSaslResponse implements ResponseFrame {
```
[the remark this answers is not in the document]
Comment (tombentley) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/frame/BareSaslResponse.java:13]: In the sense that `OpaqueRequestFrame` assumes that the frame had a standard Kafka header the answer is no, because the original support for SASL didn't use the `SaslAuthenticate` RPC at all. The client just sent the SASL bytes without any framing. _Maybe_ we could repurpose `OpaqueRequestFrame` to also handle this corner case, but it's not immediately clear to me what compromises that would come with: If we only had to check for the special case in a few places maybe that would be OK, but if many accesses required a check for the special case then I think having a separate class is cleaner. 

This comment does highlight that I've not actually implemented support for this bare request frame in the `KafkaRequestDecoder`. Over all I think I'd prefer to leave decode support as a TODO issue for now. I think the important thing is that the authentication handler is build to handle it.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/filter/NetFilter.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public interface NetFilter {
+
+    /**
+     * Determine the upstream cluster to connect to based on the information
+     * provided by the given {@code context},
+     * by invoking {@link NetFilterContext#connect(String, int, KrpcFilter[])}.
+     * @param context The context.
+     */
+    void upstreamBroker(NetFilterContext context);
```
[the remark this answers is not in the document]
Comment (tombentley) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/filter/NetFilter.java:22]: I accept that `upstreamBroker` is a poor name, but I don't think `connect` is quite right either. This method is where the proxy is asking the filter/plugin "what should I connect to?". It's up to the filter to initiate the connect by invoking `connect()` on the `NetFilterContext` passed in as an argument. What about:

* `NetFilter.upstreamBroker` → `NetFilter.selectBroker`
* `NetFilterContext.connect` → `NetFilterContext.initiateBrokerConnection`

Even then, I'm not super-convinced about using the word "broker". It assumes that the upstream server has the broker role. Maybe one day the proxy supports proxying servers which are Kraft controllers only. So maybe

* `NetFilter.upstreamBroker` → `NetFilter.selectServer`
* `NetFilterContext.connect` → `NetFilterContext.initiateConnection` (no server, because what else would you be connecting to? It can only be a server in the TCP sense).

File: pom.xml [hunk 4]
```diff
@@ -170,6 +172,12 @@
                     <version>3.10.1</version>
                     <configuration>
                         <parameters>true</parameters>
+                        <source>11</source>
+                        <target>11</target>
+<!--                        <release>11</release>-->
```
Comment (gunnarmorling) [inline pom.xml:177]: What's the end goal here? Should we not use 17 for testing? And why the change to source/target? I think I've asked about this before, but can't seem to find it. Sorry for asking again.

**Comment by gunnarmorling** [comment]
Hey @tombentley, so do we have any attack plan for getting this off the table? I feel it'd be really good to have this one in, so @k-wall can pick up. WDYT?

**Comment by tombentley** [comment]
This PR is replaced by #89
